lib/watcher.js

```diff
@@ -1,118 +1,86 @@
 
-var helper = require('./helper')
-var log = require('./logger').create('watcher')
+const helper = require('./helper')
+const log = require('./logger').create('watcher')
 
-var DIR_SEP = require('path').sep
+const DIR_SEP = require('path').sep
 
 // Get parent folder, that be watched (does not contain any special globbing character)
-var baseDirFromPattern = function (pattern) {
+function baseDirFromPattern (pattern) {
   return pattern
     .replace(/[/\\][^/\\]*\*.*$/, '')              // remove parts with *
     .replace(/[/\\][^/\\]*[!+]\(.*$/, '')         // remove parts with !(...) and +(...)
     .replace(/[/\\][^/\\]*\)\?.*$/, '') || DIR_SEP // remove parts with (...)?
 }
 
-var watchPatterns = function (patterns, watcher) {
-  // filter only unique non url patterns paths
-  var pathsToWatch = []
-  var uniqueMap = {}
-  var path
+function watchPatterns (patterns, watcher) {
+  const pathsToWatch = new Set()
 
   // expand ['a/{b,c}'] to ['a/b', 'a/c']
-  patterns = expandBraces(patterns)
-
-  patterns.forEach(function (pattern) {
-    path = baseDirFromPattern(pattern)
-    if (!uniqueMap[path]) {
-      uniqueMap[path] = true
-      pathsToWatch.push(path)
-    }
-  })
+  expandBraces(patterns)
+    .forEach(path => pathsToWatch.add(baseDirFromPattern(path)))
 
   // watch only common parents, no sub paths
-  pathsToWatch.forEach(function (path) {
-    if (!pathsToWatch.some(function (p) {
-      return p !== path && path.substr(0, p.length + 1) === p + DIR_SEP
-    })) {
+  pathsToWatch.forEach(path => {
+    if (![...pathsToWatch].some(p => p !== path && path.substr(0, p.length + 1) === p + DIR_SEP)) {
       watcher.add(path)
       log.debug('Watching "%s"', path)
     }
   })
 }
 
-// Function to test if a path should be ignored by chokidar.
-var createIgnore = function (patterns, excludes) {
+function checkAnyPathMatch (patterns, path) {
+  return patterns.some(pattern => mm(path, pattern, {dot: true}))
+}
+
+function createIgnore (patterns, excludes) {
   return function (path, stat) {
     if (!stat || stat.isDirectory()) {
       return false
     }
 
-    // Check if the path matches any of the watched patterns.
-    if (!patterns.some(function (pattern) {
-      return mm(path, pattern, {dot: true})
-    })) {
-      return true
-    }
-
-    // Check if the path matches any of the exclude patterns.
-    if (excludes.some(function (pattern) {
-      return mm(path, pattern, {dot: true})
-    })) {
-      return true
-    }
-
-    return false
+    return !checkAnyPathMatch(patterns, path) || checkAnyPathMatch(excludes, path)
   }
 }
 
-var onlyWatchedTrue = function (pattern) {
-  return pattern.watched
-}
-
-var getWatchedPatterns = function (patternObjects) {
-  return patternObjects.filter(onlyWatchedTrue).map(function (patternObject) {
-    return patternObject.pattern
-  })
+function getWatchedPatterns (patterns) {
+  return patterns
+    .filter(pattern => pattern.watched)
+    .map(pattern => pattern.pattern)
```

Comment:
while we are changing this code, lets use use either reduce, a for loop or underscore to avoid doing two iterations